### ISSUE TITLE
Workaround for Safari a bug that locks up the database

### ISSUE THREAD

**Issue by edwardxia** [dexie/Dexie.js#1030]
I discovered a bug in Safari that if I try to do `table.orderBy(index).uniqueKeys()` right after schema is created (before insert any data), Safari would throw an 'Unable to open cursor' error, after that, **every operation on this database will stuck** (even after restart the browser). It would work again, only if I delete the the database files on the disk and start over.

The workaround I found is to check if the row count is 0, before attempt to get `uniqueKeys`. For example:

``` js
if ((await db.table.count()) === 0) {
  return []
} else {
  return table.orderBy(index).uniqueKeys()
}
```

Given this bug is extremely bad as it completely locks up the db, it would be nice if Dexie can do that check internally and avoid the db to be locked up.

**Comment by edwardxia** [comment]
Here is a script to reproduce Safari's bug without Dexie:
``` js
(function (dbname) {
  const request = indexedDB.deleteDatabase(dbname)

  request.onerror = console.error

  request.onsuccess = function (event) {
    console.log(event.target.result)

    const request = indexedDB.open(dbname, 1)

    request.onupgradeneeded = function (event) {
      console.log(event.target.result)

      const db = event.target.result

      const objectStore = db.createObjectStore('customers', { keyPath: 'id' })

      objectStore.createIndex('email', 'email', { unique: false })
    }

    request.onerror = console.error
    request.onsuccess = function (event) {
      console.log(event.target.result)

      const db = event.target.result

      const index = db.transaction('customers', 'readonly').objectStore('customers').index('email')

      console.log(index)

      // const request = index.openKeyCursor() // This works
      const request = index.openKeyCursor(null, 'nextunique') // This fails and locks up

      request.onerror = console.error
      request.onsuccess = function (event) {
        console.log(event.target.result)
      }
    }
  }
})('test')
```

**Comment by dfahlander** [comment]
Thanks! This should also be filed in [Webkit bugzilla](https://bugs.webkit.org) along with the repro and the safari version. It would be something to workaround in Dexie until fixed in Safari. In which version(s) of Safari is the issue reproducable?

**Comment by edwardxia** [comment]
Version 13.1 (15609.1.20.111.8)

```
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1 Safari/605.1.15
```

**Comment by sachinrekhi** [comment]
@edwardxia I've run into a similar issue in Safari in that at some point, every IndexedDB query I run gets stuck and never resolves the promise. I don't actually do any uniqueKeys() calls, but I do have several keys() calls. Did you only run into this issue with the uniqueKeys() call?

**Comment by edwardxia** [comment]
Another issue I noticed is that if you run multiple translations in parallel, one translation will block all other translations in Safari.

I first noticed this problem in my app which uses multiple web workers to run parallel read-only transactions and I found short transactions taking excessive time only in Safari.

To prove my theory, what I did was start a very long read-only transition (a index scan on huge dataset which takes 8 seconds) in web worker, while it’s running, attempt to start another very short read-only transaction (a few milliseconds) in a different web worker. Again, in Chrome the short transaction returns immediately, but in Safari the short transaction won’t return until shortly after the long transaction returns.

I’ve seen documentation for Chrome that it’s implementation of IndexedDB support snapshot consistency and concurrent read-only transactions. I could not find any documentation on Safari around this topic. There was a WebKit issue about enable the ability of concurrent IndexedDB transactions I found, and it has been marked as fixed long time ago. Although the issue did not mention how it’s implemented, it appears that all transactions are just blocking.

@sachinrekhi And back to your question, I do still occasionally run into weird lock up situations in Safari which requires a restart of browser, and personally I think it’s related to concurrency and possibly canceling running transactions by killing web workers. If you limit it to one transaction a time, Safari would work more reliably. However, that’s a big challenge because user may open the same web app multiple time. It requires postMessage communication between different instance of the app in same browser to coordinate concurrency, and that was just too much. My project is an internal one, so that in the end I just told users if you have issues, use Chrome.

**Comment by lincolnthree** [comment]
Also noticing strange issues like this when two database read operations are run in parallel. Even from the same worker. Really loving the Safari IDB implementation.

**Comment by piotrblasiak** [comment]
Also experiencing really bad lockups like this. Does anyone have a workaround for this? Also tried to do some kind of watchdog to at least detect errors and ask the user to restart browser but it proved to be quite tricky.

**Comment by dfahlander** [comment]
I can't reproduce this in Safari Version 17.3.1 (19617.2.4.11.12). Repro link: https://jsfiddle.net/dfahlander/6rqyfag9/

**Comment by dfahlander** [comment]
Closing as the issue could be tracked down to exist in Safari 13 but solved in Safari 14 (2020)

**Comment by laukaichung** [comment]
> Closing as the issue could be tracked down to exist in Safari 13 but solved in Safari 14 (2020)

This problem still exists on my IOS devices (v 17.3.1). (Unable to open cursor)
Like OP, I have to check the total count to make sure it is safe to get uniqueKeys().

Example Reactjs code:
````
  const selectOptions = useLiveQuery(async () => {
    const values = await db.tournaments
      .orderBy("value")
      .uniqueKeys();
    return values.map(v=> ({label: xxx. value: v}))
  }, []);
```` 

````
export class MySubClassedDexie extends Dexie {
  tournaments!: Table<BracketsManagerState>;
  constructor() {
    super("myDatabase");
    this.version(1).stores({
      tournaments: "++year, value, [year+value]",
    });
  }
}
````
